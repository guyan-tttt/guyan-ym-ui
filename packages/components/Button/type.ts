import type { Component,Ref } from "vue";

// 按钮样式类型
export type ButtonType = 'primary' | 'default' | 'dashed' | 'text';

// 按钮功能类型
export type NativeType = 'button' | 'submit' | 'reset';

// 按钮大小类型
export type ButtonSize = 'large' | 'middle' | 'small';

export interface ButtonProps {
    type?: ButtonType; // 按钮类型
    size?: ButtonSize; // 按钮大小
    disabled?: boolean; // 按钮是否禁用
    loading?: boolean | { delay?: number }; // 按钮是否加载中
    plain?: boolean; // 按钮是否为朴素按钮
    nativeType?: NativeType; // 按钮原生类型
    circle?: boolean; // 按钮是否为圆形
    round?: boolean; // 按钮是否为椭圆形
    tag?: string | Component; // 按钮标签类型
    autofocus?: boolean; // 自动获取焦点
    loadingIcon?: string; // 加载图标
    useThrottle?: boolean; // 是否使用节流
    throttleDuration?: number; // 节流时间
    icon?: string; // 按钮图标
    corrugation?: boolean; // 按钮是否开启波纹效果
}

// 按钮事件
export interface ButtonEmits {
    (e: 'click', event: Event): void;
}

// 按钮组件实例
export interface ButtonInstance {
    ref:Ref<HTMLButtonElement | void>;
}

// 按钮组属性类型
export interface ButtonGroupProps {
    size?: ButtonSize;
    type?: ButtonType;
    disabled?: boolean;
}

// 按钮组上下文
export interface ButtonGroupContext {
    size?: ButtonSize;
    type?: ButtonType;
    disabled?: boolean;
}

