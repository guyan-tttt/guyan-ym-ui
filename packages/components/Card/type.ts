import type { CSSProperties } from 'vue'

type CardShadow = "always" | "hover" | "never"

export interface CardProps {
    header?: string; // 标题
    content?: string; // 内容
    footer?: string; // 底部
    shadow?: CardShadow; // 阴影显示时机
    contentStyle?:CSSProperties; // 内容自定义样式
    contentClass?: string; // 内容自定义类名
}
