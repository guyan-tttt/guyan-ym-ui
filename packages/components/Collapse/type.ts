import type { Ref } from "vue";
// 折叠面板名称
export type CollapseItemName = string | number;

// 折叠面板属性
export interface CollapseProps {
  modelValue: CollapseItemName[]; // v-model
  accordion?: boolean; // 是否手风琴模式
}
// 折叠面板子项属性
export interface CollapseItemProps {
  name: CollapseItemName; // 唯一标识
  title?: string; // 标题
  disabled?: boolean; // 是否禁用
}
// 折叠面板上下文
export interface CollapseContext {
  activeNames: Ref<CollapseItemName[]>; // 激活面板
  handleItemClick(name: CollapseItemName): void; // 点击面板
}
// 折叠面板事件
export interface CollapseEmits {
  (e: "update:modelValue", value: CollapseItemName[]): void; // 更新激活面板
  (e: "change", value: CollapseItemName[]): void; // 激活面板改变
}
