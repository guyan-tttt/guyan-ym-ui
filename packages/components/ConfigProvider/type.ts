import type { Language , TranslatePair} from '@ym-UI/locale'

export interface ConfigProviderProps {
    locale?: Language; // 语言
    extendsI18nMsg?: TranslatePair // 扩展语言包
}