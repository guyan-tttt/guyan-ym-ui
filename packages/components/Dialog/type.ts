import type { Ref } from "vue"

export interface DialogProps {
    modelValue: boolean // 是否显示
    title?: string // 标题
    width?: string | number // 宽度
    fullscreen?: boolean // 是否全屏
    top?: string // 距离顶部距离
    modal?: boolean // 是否显示遮罩
    modalClass?: string // 遮罩样式
    lockScroll?: boolean // 是否锁定滚动
    openDelay?: number // 打开延时
    closeDelay?: number // 关闭延时
    closeOnClickModal?: boolean // 点击遮罩是否关闭
    closeOnPressEscape?: boolean // 按下ESC是否关闭
    showClose?: boolean // 是否显示关闭按钮
    beforeClose?: (done: () => void) => void // 关闭前的回调
    center?: boolean // 是否居中
    closeIcon?: string // 关闭图标
    zIndex?: number // 层级
    customClass?: string // 自定义类名
    appendTo?: string // 挂载节点
    header?: boolean // 是否显示头部
    footer?: boolean // 是否显示底部
    alginCenter?: boolean // 是否居中
}

export interface DialogEmits {
    (e: 'open'): void
    (e: 'close'): void
    (e: "opened"): void
    (e: "closed"): void
    (e: "update:modelValue", value: boolean): void
}

export interface DialogSlot {
    header: () => any
    footer: () => any
    default: () => any
}

export interface DialogInstance {
    visible: Ref<boolean>
    close: () => void
    open: () => void
}