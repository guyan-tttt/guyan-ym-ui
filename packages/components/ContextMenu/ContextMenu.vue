
<script setup lang="ts">
import { ref, onMounted, watch, defineProps, useSlots, withDefaults } from "vue"
import { useContextMenu } from "./useContextMenu"
import  type { ContextMenuProps, ContextMenuOptionsItem  } from './type'

defineOptions({
  name: "YmContextMenu"
})

const props =  withDefaults(defineProps<ContextMenuProps>(),{
  type: "info",
  customize: false
})

const solts  = useSlots()


const contextMenuRef = ref<any>()

const menuRef = ref<any>()

const contextInfo = ref({
  x: 0,
  y: 0,
  visible: false
})

const menuHeight = ref(0)

const handleEnter = (el: any) => {
  el.style.height = "auto"
  menuHeight.value = el.clientHeight
  el.style.height = "0"
  requestAnimationFrame(() => {
    requestAnimationFrame(() => {
      el.style.height = menuHeight.value + "px"
    })
  })
}

// 计算鼠标是否越界
const isOut = (x: number, y: number, menu: any) => {
  const position = {
    x,
    y
  }
  if (x + menu.clientWidth > window.innerWidth) {
    position.x = window.innerWidth - 200
  }
  if (y + menu.clientHeight > window.innerHeight) {
    position.y = window.innerHeight - 200
  }
  return position
}

const openMenu = (showVisible: boolean, x: number, y: number) => {
  
  contextInfo.value.visible = showVisible
  console.log(contextInfo.value.visible,'31313');
    const position = isOut(x, y, {
      clientWidth: 150,
      clientHeight: menuHeight.value || 150
    })
    contextInfo.value.x = position.x
    contextInfo.value.y = position.y
}



onMounted(() => {
  const { x, y, visible } = useContextMenu(contextMenuRef.value)

  contextInfo.value.x = x.value
  contextInfo.value.y = y.value
  contextInfo.value.visible = visible.value

  watch(visible, () => {
    openMenu(visible.value, x.value, y.value)
  })
  watch(x, () => {
    contextInfo.value.visible = false
    setTimeout(() => {
      contextInfo.value.visible = visible.value
      const position = isOut(x.value, y.value, {
        clientWidth: 150,
        clientHeight: menuHeight.value || 150
      })
      contextInfo.value.x = position.x
      contextInfo.value.y = position.y
    }, 0)
  })
})

defineExpose({
  contextMenuRef,
  openMenu
})
</script>
<template>
    <div class="ym-content-menu" ref="contextMenuRef" >
      <slot />
      <Teleport to="body">
        <Transition @enter="handleEnter">
          <div v-if="contextInfo.visible" class="menu" ref="menuRef"  :style="{ top: contextInfo.y + 'px', left: contextInfo.x + 'px' }"
          :class="{
            [`ym-menu--${type}`]: type,
          }">
            <ul v-if="!customize">
              <li v-for="item in options" :key="item.label" @click="item.handle(item)">{{ item.label }}</li>
            </ul>
            <template v-else>
            <slot name="menu"></slot>
          </template>
          </div>
        </Transition>
      </Teleport>
    </div>
  </template>
  

  
  <style scoped >
    @import './style.css';
  </style>
  