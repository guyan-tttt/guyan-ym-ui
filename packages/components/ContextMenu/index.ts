
import ContextMenu from "./ContextMenu.vue";
import { withInstall  } from  "@ym-UI/utils"

export const YmContextMenu = withInstall(ContextMenu);

