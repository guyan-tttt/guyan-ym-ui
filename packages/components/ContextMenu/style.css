

.ym-content-menu {
    /* width: 100%;
    height: 100%; */
    display: inline-block;
  }
  .menu {
    --ym-context-menu-bg-color: var(--ym-fill-color-blank);
    --ym-collapse-header-height: 48px;
    --ym-context-menu-text-color: var(--ym-text-color-primary);
    --ym-context-menu-font-size: 13px;
    --ym-context-menu-disabled-text-color: var(--ym-disabled-text-color);
    --ym-context-menu-disabled-border-color: var(--ym-border-color-lighter);
    position: fixed;
    color: var(--ym-context-menu-text-color);
    width: 150px;
    border-radius: 20px;
    padding: 0;
    box-shadow: 0 1px 3px #ccc;
    border: 1px solid #ccc;
    transition: 0.3s ease-in;
    overflow: hidden;
    ul {
      width: 100%;
      margin: 10px 0;
      padding: 0;
      list-style: none;
      display: flex;
      flex-direction: column;
      align-items: center;
      li {
        margin: 0 10px;
        height: 30px;
        width: 100%;
        line-height: 30px;
        text-align: center;
        border-bottom: 1px solid #efefef;
        cursor: pointer;
        transition: all 0.2s ease-in;
  
        &:hover {
          background-color: #efefef;
          transform: translateY(-10px);
          box-shadow: 0 0 10px #ccc;
          border: 1px solid #fff;
          color: #46bff0;
          border-radius: 10px;
        }
        &:last-child {
          border-bottom: none;
        }
      }
    }
  }

  @each $val in primary, success, warning, info, danger {
    .ym-menu--$(val) {
      --ym-context-menu-bg-color: var(--ym-color-$(val));
      --ym-context-menu-text-color: var(--ym-color-white);
    }

  }
