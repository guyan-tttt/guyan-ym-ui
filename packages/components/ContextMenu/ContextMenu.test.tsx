import { describe, it, expect, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import ContextMenu from './ContextMenu.vue'
import { each } from 'lodash-es'
import type { ContextMenuType } from './type';

describe('ContextMenu', () => {
    // 1. 测试组件是否渲染正确
    it('renders correctly', async () => {
      const wrapper = mount(ContextMenu, {
        slots: {
          default: 'Default Slot Content',
          menu: '<p>Custom Menu Content</p>'
        },
        props: {
            customize: true
        }
      });
  
      expect(wrapper.html()).toMatchSnapshot();
    });

    // 2. 测试菜单项是否正确渲染
    it('displays the context menu with items at mouse position', async () => {
        const wrapper = mount(ContextMenu, {
          props: {
            options: [
              { label: 'Option 1', handle: vi.fn() },
              { label: 'Option 2', handle: vi.fn() },
            ]
          },
          slots: {
            default: 'Default Slot Content',
          },
          attachTo: document.body // 确保 Teleport 可以正确渲染到 body
        });
    
        // 模拟鼠标位置
        const mockMousePosition = { x: 150, y: 250 };
        await wrapper.vm.openMenu(true, mockMousePosition.x,mockMousePosition.y)
        await wrapper.vm.$nextTick();
        const menu = document.querySelector(".menu") as HTMLElement
        expect(menu).not.toBeNull() // 检查菜单是否渲染
        expect(document.body.contains(menu)).toBe(true); // 检查菜单是否渲染在 body 中
        expect(menu?.style.left).toBe(`${mockMousePosition.x}px`) // 检查菜单是否在鼠标位置
        expect(menu?.style.top).toBe(`${mockMousePosition.y}px`)
      });

      // 3.测试菜单项的样式类型
      it('has the correct style type for menu items', async () => {
       const types = ["primary","info","success","warning","danger"]
       types.forEach(async(item) => {
        console.log(item);
        
        const wrapper = mount(ContextMenu, {
          props: {
            type: item as ContextMenuType,
            options: [
              { label: 'Option 1', handle: () => {}},
              { label: 'Option 2', handle: () => {}},
            ],
          },
          slots: {
            default: 'Default Slot Content',
          },
          attachTo: document.body // 确保 Teleport 可以正确渲染到 body
        });
  
        const mockMousePosition = { x: 150, y: 250 };
        await wrapper.vm.openMenu(true, mockMousePosition.x,mockMousePosition.y)
        await wrapper.vm.$nextTick();
        const menu = document.querySelector(`.menu.ym-menu--${item}`) 
        expect(menu).not.toBeNull() 
        console.log(menu?.className);
       })
      });
})