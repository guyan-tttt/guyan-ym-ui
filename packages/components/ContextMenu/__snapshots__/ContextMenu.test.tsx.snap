// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContextMenu > renders correctly 1`] = `
"<div data-v-b7bebcf8="" class="ym-content-menu">Default Slot Content
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;
